### PR TITLE
v1alpha2 e2e initial work

### PR DESCRIPTION
**Pull request by jadarsie**
Signed-off-by: Javier Darsie jadarsie@microsoft.com

**What this PR does / why we need it**:

E2E tests initial work.

Test setup creates a management cluster (using kind) and deploys components (capi, cabpk, capz).
Test case deploys a cluster, infra cluster, machine and infra machine.

**Which issue(s) this PR fixes**:

Related to #274

**Special notes for your reviewer**:

Suggest not to spend much time reviewing the framework package. Check its `doc.go`.
See this [CAPI's issue](https://github.com/kubernetes-sigs/cluster-api/issues/1732) for context.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jadarsie** [comment]
Listing some of the pending tasks. Will move them to its own issue if/when this PR is merged. 

- Use CAPI's e2e framework instead of duplicating the module code (it references v1alpha3 types)
- Add/update e2e so it deploys multiple CP Machines
- Add/update e2e so it deploys addons
- Add/update e2e so it deploy multiple MachineDeployments
- Add/update e2e so it scales up MachineDeployments
- Add/update e2e so it scales down MachineDeployments
- Parameterize some of these variables: location, vmSize, namespace, k8sVersion, imageOffer, imagePublisher, imageSKU, imageVersion
- Parameterize CAPI/CABPK versions
- Update ci-e2e so it deletes all dangling resources

File: test/e2e/azure_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+	Expect(infrav1.AddToScheme(scheme)).To(Succeed())
+
+	managerImage, found := os.LookupEnv("MANAGER_IMAGE")
+	Expect(found).To(BeTrue(), fmt.Sprint("MANAGER_IMAGE not set"))
+
+	mgmt, err = kind.NewCluster(ctx, "mgmt", scheme, managerImage)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(mgmt).NotTo(BeNil())
+
+	// TODO Figure out how to keep these versions in sync across the code base
```
Comment (CecileRobertMichon) [inline test/e2e/azure_suite_test.go]: can you open an issue to track this please?

File: test/e2e/azure_test.go [hunk 2]
```diff
@@ -16,228 +18,275 @@ limitations under the License.
...
+					CreateTimeout: 60 * time.Minute,
+				})
+				framework.CleanUp(&framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				})
+			})
+		})
+
+		// TODO: Deploy multiple Control Plane
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go]: let's open good first issues for each of these

File: test/e2e/azure_test.go [hunk 3]
```diff
@@ -16,228 +18,275 @@ limitations under the License.
...
+	}
+	return framework.Node{
+		Machine:         machine,
+		InfraMachine:    infraMachine,
+		BootstrapConfig: bootstrapConfig,
+	}
+}
+
+func cloudConfig(cn string) string {
+	// TODO This is ugly
```
Comment (CecileRobertMichon) [inline test/e2e/azure_test.go]: lol

File: test/e2e/framework/clean_up.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+// CleanUpInput are all the dependencies needed to clean up a Cluster API cluster.
+type CleanUpInput struct {
+	Management    *kind.Cluster
+	Cluster       *clusterv1.Cluster
+	DeleteTimeout time.Duration
+}
+
+func (c *CleanUpInput) setDefaults() {
+	if c.DeleteTimeout == 0*time.Second {
```
Comment (CecileRobertMichon) [inline test/e2e/framework/clean_up.go]: nit: spaces missing

File: test/e2e/framework/clean_up.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	ctx := context.Background()
+	gomega.Expect(mgmtClient.Delete(ctx, input.Cluster)).NotTo(gomega.HaveOccurred())
+
+	gomega.Eventually(func() []clusterv1.Cluster {
+		clusters := clusterv1.ClusterList{}
+		c, err := input.Management.GetClient()
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(c.List(ctx, &clusters)).NotTo(gomega.HaveOccurred())
+		return clusters.Items
+	}, input.DeleteTimeout, 20*time.Second).Should(gomega.HaveLen(0))
```
Comment (CecileRobertMichon) [inline test/e2e/framework/clean_up.go]: maybe we should also check that the infra is deleted in the future

File: test/e2e/framework/clean_up.go (hunk 4 above)
Comment (jadarsie) [reply test/e2e/framework/clean_up.go]: this `0*time.Second` is made by the go tooling, if that's what you mean.

File: test/e2e/azure_suite_test.go (hunk 1 above)
Comment (jadarsie) [reply test/e2e/azure_suite_test.go]: added [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/339#issuecomment-564200720)

File: test/e2e/azure_test.go (hunk 2 above)
Comment (jadarsie) [reply test/e2e/azure_test.go]: listed [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/339#issuecomment-564200720)

File: test/e2e/framework/clean_up.go (hunk 5 above)
Comment (jadarsie) [reply test/e2e/framework/clean_up.go]: added to the list above

File: test/e2e/azure_suite_test.go (hunk 1 above)
Comment (jadarsie) [reply test/e2e/azure_suite_test.go]: added to the list above

**Comment by justaugustus** [comment]
/hold
@jadarsie -- We should create the umbrella issue for follow-up tasks _before_ merge.
Please open that and link it here before we approve.

**Comment by jadarsie** [comment]
> /hold
> @jadarsie -- We should create the umbrella issue for follow-up tasks _before_ merge.
> Please open that and link it here before we approve.

FYI, created a tracking issue, will create a set of more granular issues as soon as I can

**Comment by CecileRobertMichon** [comment]
Thanks for opening issue #344 @jadarsie! 

/approve
/hold cancel

**Pull request merged**